jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedGraph.java

```diff
@@ -17,111 +17,82 @@
 
-    private final Map<E, Double> weightMap;
+    private static final long serialVersionUID = -6838132233557L;
+    private final Map<E, Double> weights;
 
     /**
-     * Constructor for AsWeightedGraph.
+     * Constructor for AsWeightedGraph
      *
-     * @param g the backing graph over which a weighted view is to be created.
-     * @param weightMap A mapping of edges to weights. If an edge is not present in the weight map,
-     *        the edge weight for the underlying graph is returned. Note that a live reference to
-     *        this map is retained, so if the caller changes the map after construction, the changes
-     *        will affect the AsWeightedGraph instance as well.
-     */
-    public AsWeightedGraph(Graph<V, E> g, Map<E, Double> weightMap)
-    {
-        super(g);
-        this.weightMap = Objects.requireNonNull(weightMap, "Weight map cannot be null");
-    }
-
-    /**
-     * {@inheritDoc}
+     * @param graph   the backing graph over which an weighted view is to be created.
+     * @param weights the map containing the edge weights
+     * @throws NullPointerException if the graph is null
      */
-    @Override
-    public void setEdgeWeight(E e, double weight)
+    public AsWeightedGraph(Graph<V, E> graph, Map<E, Double> weights)
     {
-        if (super.getType().isWeighted()) {
-            super.setEdgeWeight(e, weight);
-        }
-
-        // Always modify the weight map. It would be a terrible violation
-        // of the use contract to silently ignore changes to the weights.
-        weightMap.put(e, weight);
+        super(graph);
+        this.weights = weights;
     }
 
     /**
-     * {@inheritDoc}
+     * Returns the weight assigned to a given edge.
+     * If there is no edge weight set for the given edge, the value of the backing graph's
+     * getEdgeWeight method is returned.
+     *
+     * @param e edge of interest
+     * @return the edge weight
+     * @throws NullPointerException if the edge is null
      */
-    @Override
-    public double getEdgeWeight(E e)
+    @Override public double getEdgeWeight(E e)
     {
-        double weight;
+        Objects.requireNonNull(e);
+        Double weight = this.weights.get(e);
 
-        // Always return the value from the weight map first and
-        // only pass the call through as a backup
-        if (weightMap.containsKey(e)) {
-            weight = weightMap.get(e);
-        } else {
+        if (Objects.isNull(weight)) {
             weight = super.getEdgeWeight(e);
         }
 
         return weight;
     }
 
     /**
-     * {@inheritDoc}
+     * Assigns a weight to an edge.
+     *
+     * @param e      edge on which to set weight
+     * @param weight new weight for edge
+     * @throws NullPointerException if the edge is null
      */
-    @Override
-    public GraphType getType()
+    @Override public void setEdgeWeight(E e, double weight)
+    {
+        Objects.requireNonNull(e);
+        this.weights.put(e, weight);
```

Comment:
Not sure whether we should also change the weight of the backing graph, if applicable. I think our old contract did that. @jsichi @jkinable  Opinions?

Reply:
Both behaviors seem useful depending on the application; how about add a boolean constructor parameter for writeThrough, and make it true by default?

Reply:
According to [java by comparison](http://java.by-comparison.com/), one should use an ENUM here. Maybe `LAYER_ONLY` and `MODIFY_BACKING_GRAPH` to make the enum really distinguished?

Reply:
If clarity is the aim, then use WRITE_THROUGH vs NO_WRITE_THROUGH.  If I see LAYER_ONLY, I'll have no idea what that means unless I go read the corresponding enum comment anyway.  But in my opinion, a boolean works just as well here since we're unlikely to add more parameters.